src/drivers/magnetometer/memsic/mmc5616/mmc5616.cpp

```diff
@@ -0,0 +1,340 @@
+/**
+ * @file mmc5616.cpp
+ *
+ * Implementation of Memsic MMC5616 Driver
+ */
+
+#include "mmc5616.hpp"
+
+// define to a noop for quiet
+#define MEMSIC_INFO(...)        {PX4_INFO(__VA_ARGS__);}
+
+using namespace time_literals;
+
+static constexpr int16_t combine(uint8_t msb, uint8_t lsb)
+{
+  return (msb << 8u) | lsb;
+}
+
+MMC5616::MMC5616(device::Device *interface, const I2CSPIDriverConfig &config) :
+	I2CSPIDriver(config),
+	_interface(interface),
+	_px4_mag(interface->get_device_id(), config.rotation)
+{
+}
+
+MMC5616::~MMC5616()
+{
+  perf_free(_comms_errors);
+  perf_free(_reset_perf);
+  perf_free(_timeout_perf);
+  delete _interface;
+}
+
+int MMC5616::init()
+{
+  return Reset() ? 0 : -1;
+}
+
+bool MMC5616::Reset()
+{
+  _state = STATE::RESET;
+  ScheduleClear();
+  ScheduleNow();
+  return true;
+}
+
+void MMC5616::print_status()
+{
+  I2CSPIDriverBase::print_status();
+  perf_print_counter(_reset_perf);
+  perf_print_counter(_timeout_perf);
+}
+
+bool MMC5616::checkIDs()
+{
+  const uint8_t chip_id = RegisterRead(Register::R_CHIP_ID);
+  const uint8_t product_id = RegisterRead(Register::R_PRODUCT_ID);
+
+  if (chip_id != EXPECTED_CHIP_ID)
+  {
+    PX4_DEBUG("unexpected Chip ID: 0x%02x (0x%02x)", chip_id, EXPECTED_CHIP_ID);
+  }
+
+  if (product_id != EXPECTED_PRODUCT_ID)
+  {
+    PX4_DEBUG("unexpected Product ID: 0x%02x (0x%02x)", product_id, EXPECTED_PRODUCT_ID);
+    return false;
+  }
+
+  PX4_DEBUG("Detected device: ChipID 0x%02x, ProductID 0x%02x", chip_id, product_id);
+
+  return true;
+}
+
+int MMC5616::probe()
+{
+  for (uint8_t i = 0; i < 3; i++)
+  {
+    bool check_ids = checkIDs();
+    if (check_ids)
+    {
+      return PX4_OK;
+    }
+    usleep(100);
+  }
```

Comment:
The retries here are unecessary, I would remove this loop. If the device has a power-on time then the delay should happen as a state with rescheduling. We can't use usleep in drivers like this since they share the same work queue with other I2C devices on the same bus. Sleeping in the work queue blocks the entire queue.